tools/functional-tester/etcd-runner/command/watch_command.go

```diff
@@ -49,154 +42,23 @@ func runWatcherFunc(cmd *cobra.Command, args []string) {
+		TotalKeys:        totalKeys,
 	}
-	revision = gr.Header.Revision
-
-	ctxt, cancel := context.WithDeadline(ctx, time.Now().Add(runningTime*time.Second))
-	defer cancel()
-
-	// generate and put keys in cluster
-	limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
-
-	go func() {
-		for _, key := range keys {
-			for _, prefix := range prefixes {
-				if err = limiter.Wait(ctxt); err != nil {
-					return
-				}
-				if err = putKeyAtMostOnce(ctxt, client, roundPrefix+"-"+prefix+"-"+key); err != nil {
-					log.Fatalf("failed to put key: %v", err)
-					return
-				}
-			}
-		}
-	}()
-
-	ctxc, cancelc := context.WithCancel(ctx)
 
-	wcs := make([]clientv3.WatchChan, 0)
-	rcs := make([]*clientv3.Client, 0)
-
-	for _, prefix := range prefixes {
-		for j := 0; j < watchPerPrefix; j++ {
-			rc := newClient(eps, dialTimeout)
-			rcs = append(rcs, rc)
-
-			watchPrefix := roundPrefix + "-" + prefix
-
-			wc := rc.Watch(ctxc, watchPrefix, clientv3.WithPrefix(), clientv3.WithRev(revision))
-			wcs = append(wcs, wc)
-
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				checkWatchResponse(wc, watchPrefix, keys)
-			}()
-		}
-	}
-	wg.Wait()
-
-	cancelc()
-
-	// verify all watch channels are closed
-	for e, wc := range wcs {
-		if _, ok := <-wc; ok {
-			log.Fatalf("expected wc to be closed, but received %v", e)
-		}
-	}
-
-	for _, rc := range rcs {
-		rc.Close()
-	}
-
-	if err = deletePrefix(ctx, client, roundPrefix); err != nil {
-		log.Fatalf("failed to clean up keys after test: %v", err)
-	}
-}
-
-func checkWatchResponse(wc clientv3.WatchChan, prefix string, keys []string) {
-	for n := 0; n < len(keys); {
-		wr, more := <-wc
-		if !more {
-			log.Fatalf("expect more keys (received %d/%d) for %s", len(keys), n, prefix)
-		}
-		for _, event := range wr.Events {
-			expectedKey := prefix + "-" + keys[n]
-			receivedKey := string(event.Kv.Key)
-			if expectedKey != receivedKey {
-				log.Fatalf("expected key %q, got %q for prefix : %q\n", expectedKey, receivedKey, prefix)
-			}
-			n++
-		}
-	}
-}
-
-func putKeyAtMostOnce(ctx context.Context, client *clientv3.Client, key string) error {
-	gr, err := getKey(ctx, client, key)
-	if err != nil {
-		return err
-	}
-
-	var modrev int64
-	if len(gr.Kvs) > 0 {
-		modrev = gr.Kvs[0].ModRevision
-	}
-
-	for ctx.Err() == nil {
-		_, err := client.Txn(ctx).If(clientv3.Compare(clientv3.ModRevision(key), "=", modrev)).Then(clientv3.OpPut(key, key)).Commit()
-
-		if err == nil {
-			return nil
-		}
-	}
-
-	return ctx.Err()
-}
-
-func deletePrefix(ctx context.Context, client *clientv3.Client, key string) error {
-	for ctx.Err() == nil {
-		if _, err := client.Delete(ctx, key, clientv3.WithPrefix()); err == nil {
-			return nil
-		}
-	}
-	return ctx.Err()
-}
-
-func getKey(ctx context.Context, client *clientv3.Client, key string) (*clientv3.GetResponse, error) {
-	for ctx.Err() == nil {
-		if gr, err := client.Get(ctx, key); err == nil {
-			return gr, nil
-		}
-	}
-	return nil, ctx.Err()
+	runner.RunWatch(wcf)
```

Comment:
i have the error checking at https://github.com/fanminshi/etcd/blob/e84ab64c17f22fc8986cd919223a27f036023a39/tools/functional-tester/etcd-runner/runner/watch.go#L37

But not sure if that's the expected behavior for runner command.